pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -442,24 +449,31 @@ func (p *staticPolicy) guaranteedCPUs(pod *v1.Pod, container *v1.Container) int
 }
 
 func (p *staticPolicy) podGuaranteedCPUs(pod *v1.Pod) int {
+	// The sum of requested CPUs by app containers.
+	requestedByAppContainers := 0
+	for _, container := range pod.Spec.Containers {
+		if _, ok := container.Resources.Requests[v1.ResourceCPU]; !ok {
+			continue
+		}
+		requestedByAppContainers += p.guaranteedCPUs(pod, &container)
+	}
+
 	// The maximum of requested CPUs by init containers.
 	requestedByInitContainers := 0
+	requestedByRestartableInitContainers := 0
 	for _, container := range pod.Spec.InitContainers {
 		if _, ok := container.Resources.Requests[v1.ResourceCPU]; !ok {
 			continue
 		}
 		requestedCPU := p.guaranteedCPUs(pod, &container)
-		if requestedCPU > requestedByInitContainers {
-			requestedByInitContainers = requestedCPU
-		}
-	}
-	// The sum of requested CPUs by app containers.
-	requestedByAppContainers := 0
-	for _, container := range pod.Spec.Containers {
-		if _, ok := container.Resources.Requests[v1.ResourceCPU]; !ok {
-			continue
+		if types.IsRestartableInitContainer(&container) {
+			requestedByAppContainers += requestedCPU
+			requestedByRestartableInitContainers += requestedCPU
+		} else {
+			if requestedByRestartableInitContainers+requestedCPU > requestedByInitContainers {
+				requestedByInitContainers = requestedByRestartableInitContainers + requestedCPU
+			}
```

Comment:
I'm not sure about this part. The intent of the previous logic was something along these lines:

1. init container execute one of the another, so we can reuse resources. So let's take the shape of the biggest container, and take that as footprint of the pod for CPU allocation purposes
2. the above can be extended to app containers: init container will be all terminated by the time app container start executing, so let's consider the shape of the app containers when looking for the biggest.

The net result of all the above could mean temporary overallocation, but the assumption is (was) that init containers execute for a time vastly inferior to app container.

Because all of the above
1. I'm not clear ehy we need to take into account restartableinitcontainers when we look for the shape of init containers, especially since we (correctly) account them against the shape of app containers right above
2. I would prefer to not conflate app container and restartable (sidecar) init container up until the very end, or at least introduce a new cumulative variable like `requestedByLongRunningContainers = requestedByAppContainers + requestedByRestartableInitContainers`.

Reply:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/753-sidecar-containers#resources-calculation-for-scheduling-and-pod-admission

^ This is the new resource calculation after the SidecarContainers feature.

> 1. I'm not clear ehy we need to take into account restartableinitcontainers when we look for the shape of init containers, especially since we (correctly) account them against the shape of app containers right above

> InitContainerUse(i) = Sum(sidecar containers with index < i) + InitContainer(I)

it should be considered with init containers as well.

e.g.
```yaml
initContainers:
- name: restartable-init-1
  resources:
    requests:
      cpu: 1
- name: init-2
  resources:
    requests:
      cpu: 2
containers:
- name: app-3
  resources:
    requests:
      cpu: 1
```

the maximum CPU should be 3(1+2) not 2(1+1).

> 2\. I would prefer to not conflate app container and restartable (sidecar) init container up until the very end, or at least introduce a new cumulative variable like `requestedByLongRunningContainers = requestedByAppContainers + requestedByRestartableInitContainers`.

I'd fix it for clarity.

Reply:
fixed it.